numpy/core/function_base.py

```diff
@@ -123,25 +123,47 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
     if num < 0:
         raise ValueError("Number of samples, %s, must be non-negative." % num)
     div = (num - 1) if endpoint else num
+    # Make sure one can use variables that have an __array_interface__, gh-6634
+    start = asanyarray(start)
+    stop  = asanyarray(stop)
+
+    # check if dealing with datetime64 or timedelta64
+    is_datelike = start.dtype.kind in 'Mm'
+    if is_datelike:
+        # For datetime/timedelta, dtype must be specified
+        # start and stop will first be cast to this dtype.
+        # Warning: dtype must have high enough precision so
+        # that it is at all possible to draw num equispaced
+        # samples.
+        assert dtype is not None, "linspace with {} objects cannot be called "\
+            "with dtype=None".format(start.dtype)
+        start = start.astype(dtype)
+        stop = stop.astype(dtype)
+        # y cannot be cast to datetime64/timedelta64, but left
+        # as int64, since later on we will do y * delta, and
+        # <m8[s] * <m8[s] is not allowed.
+        delta = stop - start
+        y = _nx.arange(0, num).reshape((-1,) + (1,) * ndim(delta))
 
-    # Convert float/complex array scalars to float, gh-3504
-    # and make sure one can use variables that have an __array_interface__, gh-6634
-    start = asanyarray(start) * 1.0
-    stop  = asanyarray(stop)  * 1.0
+    else:
+        # Convert float/complex array scalars to float, gh-3504
+        start = start * 1.0
+        stop  = stop  * 1.0
+        dt = result_type(start, stop, float(num))
+        delta = stop - start
+        y = _nx.arange(0, num, dtype=dt).reshape((-1,) + (1,) * ndim(delta))
 
-    dt = result_type(start, stop, float(num))
-    if dtype is None:
-        dtype = dt
+        if dtype is None:
+            dtype = dt
 
-    delta = stop - start
-    y = _nx.arange(0, num, dtype=dt).reshape((-1,) + (1,) * ndim(delta))
     # In-place multiplication y *= delta/div is faster, but prevents the multiplicant
     # from overriding what class is produced, and thus prevents, e.g. use of Quantities,
-    # see gh-7142. Hence, we multiply in place only for standard scalar types.
-    _mult_inplace = _nx.isscalar(delta)
+    # see gh-7142. Hence, we multiply in place only for standard scalar types,
+    # excepting datetime64 and timedelta64
+    _mult_inplace = _nx.isscalar(delta) and (not is_datelike)
     if num > 1:
         step = delta / div
-        if _nx.any(step == 0):
+        if (not is_datelike) and _nx.any(step == 0):
```

Comment:
I think this would be better as `if np.issubdtype(step.dtype, np.floating)`, since that's the only case where denormal numbers apply

Reply:
I tried that, and it fails `test_equivalent_to_arange`:
```python
    def test_equivalent_to_arange(self):
        for j in range(1000):
            assert_equal(linspace(0, j, j+1, dtype=int),
>                        arange(j+1, dtype=int))
E           AssertionError: 
E           Arrays are not equal
E           
E           Mismatched elements: 1 / 23 (4.35%)
E           Max absolute difference: 1
E           Max relative difference: 0.06666667
E            x: array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 14, 16,
E                  17, 18, 19, 20, 21, 22])
E            y: array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16,
E                  17, 18, 19, 20, 21, 22])
```
In the last line of `linspace`, the y[15] happens to be `14.999999999999998`, which becomes 14 when cast to `int` in the test.

Reply:
Hmmm, interesting. Although I wonder if it is generally true that doing the division first is more precise. But it is slower.

Reply:
This line of code is no longer present in the PR any more, as I refactored it to use `ndarray.view`